k8sClient.py

```diff
@@ -11,7 +12,16 @@
 
 
 class K8sClient:
-    def __init__(self, kubeconfig: str):
```

Comment:
Let's add a new function called

```
    @staticmethod
    def withDetectedKubeConfig(kubeconfig: Optional[str]) -> str:
        if not kubeconfig:
            kubeconfig = os.environ.get("KUBECONFIG")
            if not kubeconfig:
                raise ValueError("No kubeconfig given and KUBECONFIG environment not set")
        return kubeconfig
```

And keep the mandatory argument to K8sClient. You know what I think about `Optional` args which are then followed by `if .. is None: get default`.

Reply:
> Let's add a new function called

I don't know where you looking at (github shows so little context).

This `withDetectedKubeConfig()` is exactly the same as the `detect_kubeconfig()` function that the patch adds. Do you mean, `detect_kubeconfig()` should be renamed to `withDetectedKubeConfig()`? OK. I don't think that's a better name, but whatever.

---

> And keep the mandatory argument to K8sClient.

That does not solve the problem.

The problem that the patch solves, is that you can run

```bash
      $ KUBECONFIG=... \
        CDA_LOG_LEVEL=debug \
        python -c 'if 1:
            import extraConfigSriov as e
            repo_dir = "/root/sriov-network-operator"
            e._sno_repo_setup(repo_dir, repo_wipe=False)
            e._sno_make_deploy(repo_dir, build_local=True)
        '
```

for manual testing or manually running the build command.

> You know what I think about Optional args which are then followed by if .. is None: get default.

Yes.

Reply:
To clarify, I'm suggesting a factory pattern here.

Reply:
```
class K8sClient:
    def __init__(self, kubeconfig: str):
        self._kc = kubeconfig
        with open(kubeconfig) as f:
            c = yaml.safe_load(f)
        self._api_client = kubernetes.config.new_client_from_config_dict(c)
        self._client = kubernetes.client.CoreV1Api(self._api_client)

    @staticmethod
    def with_env_kubeconfig() -> str:
        if not kubeconfig:
            kubeconfig = os.environ.get("KUBECONFIG")
            if not kubeconfig:
                raise ValueError("KUBECONFIG environment is not set")
        return K8sClient(kubeconfig)

```

Reply:
I guess, then `_sno_make_deploy()` also must not have an `kubeconfig: Optional[str]=None` argument?

With an optional argument, you can have a top-level parameter `kubeconfig: Optional[str]`, which gets passed down, and the appropriate places will create a suitable `K8sConfig` out of that. If you require that there are no optionals all the way down, then at the highest level, the value must be fetched from the environment already. It also leaves no place to call `with_env_kubeconfig()` and makes it pointless.

---

On the other hand, if you allow that `_sno_make_deploy()` can have a parameter `kubeconfig: Optional[str] = None`, then it must itself do:
```
     if kubeconfig is None:
          client = k8sClient.K8sClient.with_env_kubeconfig()
    else:
          client = k8sClient.K8sClient(kubeconfig)
```
(and every other place that would allow for fallback to `$KUBECONFIG`  would have to reimplement those lines).

---

I drop the patch.

Reply:
exactly. Imo it's better to depend on the env in 1 place (top level) and remove all the copy-paste `if something is in env` code from everywhere including all the `Optional[...]` args.

Reply:
That takes away the reason why this was added. To be able to run the code conveniently form the command line, without calling all the code from the highest level down.

the patch is dropped.